addons/bus/tests/test_websocket_caryall.py

```diff
@@ -0,0 +1,131 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import json
+import time
+from collections import defaultdict
+from unittest.mock import patch
+
+from odoo.api import Environment
+from odoo.tests import common, new_test_user
+from .common import WebsocketCase
+from ..models.bus import dispatch
+from ..websocket import (
+    CloseCode,
+    Frame,
+    Opcode,
+    TimeoutManager,
+    TimeoutReason,
+    Websocket
+)
+
+@common.tagged('post_install', '-at_install')
+class TestWebsocketCaryall(WebsocketCase):
+    def test_lifecycle_hooks(self):
+        events = []
+        with patch.object(Websocket, '_event_callbacks', defaultdict(set)):
+            @Websocket.onopen
+            def onopen(env, websocket):  # pylint: disable=unused-variable
+                self.assertIsInstance(env, Environment)
+                self.assertIsInstance(websocket, Websocket)
+                events.append('open')
+
+            @Websocket.onclose
+            def onclose(env, websocket):  # pylint: disable=unused-variable
+                self.assertIsInstance(env, Environment)
+                self.assertIsInstance(websocket, Websocket)
+                events.append('close')
+
+            ws = self.websocket_connect()
+            ws.close(CloseCode.CLEAN)
+            self.wait_remaining_websocket_connections()
+            self.assertEqual(['open', 'close'], events)
+
+    def test_instances_set(self):
+        first_ws = self.websocket_connect()
+        second_ws = self.websocket_connect()
+        self.assertEqual(len(Websocket._instances), 2)
+        first_ws.close(CloseCode.CLEAN)
+        second_ws.close(CloseCode.CLEAN)
+        self.wait_remaining_websocket_connections()
+        self.assertEqual(len(Websocket._instances), 0)
+
+    def test_timeout_manager_no_response_timeout(self):
+        with patch.object(TimeoutManager, 'TIMEOUT', 10):
+            timeout_manager = TimeoutManager()
+            # A PING frame was just sent, if no pong has been received
+            # within 10 seconds, the connection should have timed out.
+            timeout_manager.acknowledge_frame_sent(Frame(Opcode.PING))
+            self.assertEqual(timeout_manager._awaited_opcode, Opcode.PONG)
+            with patch('time.time', return_value=time.time() + 5):
```

Comment:
The latest diff doesn't `assertFalse(has_timed_out())` before the `tick(default=10s)` anymore. Was it done on purpose? It was nice to read the assert no timeout => tick() => assert timeout succession.

Reply:
Yes, I thought it was not adding much since we would figure if the connection kept timing out. But reading your comment I realize it was not a good idea, I'll just add it again.